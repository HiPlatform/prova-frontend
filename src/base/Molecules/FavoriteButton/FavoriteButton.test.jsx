import React from 'react';
import { shallow } from 'enzyme';
import { partial } from 'ramda';

import { FavoriteButton } from './index';
import toggleFavorite from './FavoriteButton.actions';
import FavoriteReducer from './FavoriteButton.reducer';

describe('FavoriteButton', () => {
  const initial = {
    artist: { items: [] },
    album: { items: [] },
    track: { items: [] },
  };
  const fnToggle = jest.fn();
  const props = {
    fnToggle,
  };
  const component = shallow(<FavoriteButton {...props} />);
  const randomAction = { type: 'random' };

  it('should render without crashing', () => {
    expect(component).toMatchSnapshot();
  });

  it('should call the toggle function when heart is clicked', () => {
    component.find('button').simulate('click');
    expect(fnToggle).toHaveBeenCalled();
  });

  describe('[Actions] FavoriteButton >', () => {
    it('toggleFavorite', () => {
      const item = 'something';
      const list = 'artist';
      const toggleAction = toggleFavorite(list, item);
      const expectedResult = { type: 'fav', item, list };

      expect(toggleAction).toEqual(expectedResult);
    });
  });

  describe('[Reducer] FavoriteButton >', () => {
    const reducer = partial(FavoriteReducer, [initial]);

    it('has default called', () => {
      expect(reducer(randomAction)).toEqual(initial);
    });

    it('puts the item in the favorites array and removes on a second call', () => {
      const item = 'something';
      const list = 'artist';
      const toggleAction = toggleFavorite(list, item);
      const expectedResult = { ...initial, artist: { items: [item] } };

      expect(reducer(toggleAction)).toEqual(expectedResult);
    });

    it('removes the item of the favorites array', () => {
      const item = 'something';
      const list = 'artist';
      const newReducer = partial(FavoriteReducer, [{ ...initial, artist: { items: [item] } }]);
      const toggleAction = toggleFavorite(list, item);

      expect(newReducer(toggleAction)).toEqual(initial);
    });
  });
});
