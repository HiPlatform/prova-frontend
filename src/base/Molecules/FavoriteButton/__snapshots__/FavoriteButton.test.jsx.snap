// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteButton should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    favorites={Array []}
    fnToggle={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        className="fa fa-heart"
/>,
      "className": "favoriteButton ",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "fa fa-heart",
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          className="fa fa-heart"
/>,
        "className": "favoriteButton ",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fa fa-heart",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
